### ISSUE TITLE
[memory] make session digests semantically structured so they can be segmented and indexed

### ISSUE THREAD

**Issue by lucasrp** [lucasrp/edge-of-chaos#334]
## Problem

If we add session digests to preflight, a free-form prose summary is not enough.

A good summary for a human is still weak for the system if it cannot be:
- segmented cleanly
- diffed incrementally
- marked as resolved / superseded / contradicted
- indexed for retrieval later

If we only generate a blob of prose, we will get:
- decent short-term readability
- poor retrieval quality
- poor incremental redigest behavior
- poor ability to separate active vs stale session memory

## Proposal

Design session digests as **semantic objects**, not just summaries.

The digest should be structured from the start so it can support both:
1. hot preflight injection
2. later segmentation / indexing / retrieval

## Why this matters

We want two different uses from the same memory layer:
- a compact recent-session digest injected into substantive skill preflight
- periodic full redigests that become retrievable artifacts in the corpus

That only works well if the digest is architected into semantic sections/items instead of one monolithic text block.

## Suggested schema shape

A digest should contain explicit structured sections such as:
- `operator_intent`
- `active_focus`
- `recent_decisions`
- `unresolved_questions`
- `repeated_failures`
- `contradictions`
- `resolved_items`
- `superseded_items`
- `entities`
- `skills_involved`
- `decision_types`
- `failure_types`
- `status_transitions`

Each item should be small and structured, for example:
- short text
- evidence/source pointer
- recency
- status
- tags/entities

## Retrieval implications

Session digests do **not** have strong thematic pertinence by default.

So they should not be indexed like ordinary thematic documents.

Instead, periodic redigests should be segmented/indexed by:
- entities mentioned
- decisions
- unresolved questions
- failures / loops
- contradictions
- skills involved
- recency
- semantic status (`active`, `resolved`, `superseded`, `contradicted`)

That allows retrieval to use them as structured episodic memory instead of prose blobs.

## Lifecycle model

Recommended split:
- incremental digest: hot, operational, not indexed
- periodic full redigest: canonical, segmented, indexable

So the periodic redigest should already be emitted in a shape that can be atomized into retrievable units.

## Success criteria

- session digest schema supports clean incremental updates
- periodic redigests can be segmented into retrieval units without reparsing prose heuristically
- retrieval can rank session memory by entities, decisions, failures, recency, and semantic status
- preflight can inject a compact digest while the same memory layer still supports deeper recall later

**Comment by lucasrp** [comment]
A stronger framing for this digest layer:

This should not be treated as just "recent conversation memory". It can become the best feedback surface the system has, because it carries a much more direct signal of **required state change** than the agent's own guesswork.

So the digest / redigest should be used as a kind of **operator pressure ledger** over the system:
- what the operator keeps asking to change
- what keeps being corrected repeatedly
- what still has not been crystallized into workflow / procedure / capability / policy
- which desired state changes are still open

That suggests additional structured sections such as:
- `required_state_changes`
- `repeated_operator_guidance`
- `recent_corrections`
- `workflow_candidates`
- `capability_candidates`
- `resolved_promotions`
- `ignored_or_repeated_failures`

Each item should carry at least:
- short text
- target (`workflow|procedure|capability|policy|skill`)
- recency
- repeat count
- evidence/source pointer
- status (`active|promoted|resolved|superseded|ignored`)

This should also have **direct consumers**, not just be readable context:
1. preflight / skill request injection
2. workflow recommendation / promotion logic
3. health / learning metrics

In other words: use this as a canonical feedback object about the delta between current system behavior and desired operator-imposed state, not just as a memory summary.

**Comment by lucasrp** [comment]
## Reframe before locking the schema

The 13 sections in the original proposal conflate three concerns that production systems in 2026 keep separated:

1. **Canonical storage** (atomic, append-only, bitemporal) — small and indexable
2. **Narrative grouping** (episodes / arcs) — preserves trajectory, not just facts
3. **Operator-facing render** (template with N fixed sections) — LLM synthesis over (1) and (2), with provenance back to atoms

The item-level metadata in the issue (status, evidence pointer, recency, tags/entities) is exactly right — keep it on the atom. The 13 sections, however, mix render outputs (which need LLM aggregation for things like `repeated_failures`, `contradictions`, `status_transitions`) with storage units. Better to have one atom schema and let render templates evolve.

### Convergent pattern from the market (2026)

| System | Storage unit | Status handling | Retrieval ranking |
|---|---|---|---|
| SEEM (arxiv 2601.06411) | Episodic Event Frame + Graph quadruples | Bitemporal | Provenance expansion + relational propagation |
| Mem0 | Atomic statement | Supersede on contradiction | Recency + relevance + salience |
| Graphiti | Fact triple | Append-only, invalidate without delete | Temporal validity windows |
| agentpatterns (coding agents) | Episode (trigger/context/attempts/resolution/lesson) | Outcome boundary | Trigger + outcome + recency |
| Claude Code v2.0.64+ (native) | session-memory/summary.md (markdown blob) | None explicit | None — re-injected wholesale |

Across systems: storage unit is small and atomic; render is separate; status is bitemporal; salience is the third retrieval dim.

### Atom schema (canonical layer)

```json
{
  "id":          "atom_<12 hex>",
  "kind":        "intent | decision | question | failure | contradiction | entity_state | resolution",
  "content":     "1-3 sentences, no markdown",
  "status":      "active | resolved | superseded | contradicted",
  "valid_from":  "<ISO datetime>",
  "valid_until": "<ISO datetime | null>",
  "supersedes":  ["atom_id_old"],
  "entities":    [{"name": "...", "type": "skill|tool|file|concept|person|issue|session"}],
  "salience":    0.0,
  "provenance":  {"source_kind": "session", "source_id": "<sid>", "msg_range": [start, end]},
  "created_at":  "<ISO datetime>",
  "embedding_id": "<fk | null>"
}
```

### Episode schema (narrative arcs over atoms)

```json
{
  "id":         "ep_<12 hex>",
  "trigger":    "what initiated this arc",
  "context":    "state at episode start",
  "atom_ids":   ["atom_a", "atom_b", "..."],
  "outcome":    "resolved | abandoned | deferred | in_progress",
  "lesson":     "abstracted takeaway, instance-stripped",
  "salience":   0.0,
  "entities":   ["..."],
  "created_at": "<ISO datetime>"
}
```

### Retrieval score (preflight injection)

```python
score(atom, q) = (
    W_STATUS   * STATUS_WEIGHT[atom.status]            # active=1.0, resolved=0.3, superseded=0.0
  + W_SALIENCE * atom.salience
  + W_RECENCY  * exp(-age_days / HALF_LIFE_DAYS)       # 14d default
  + W_SEMANTIC * cosine(atom.embedding, q.embedding)
  + W_ENTITY   * jaccard(atom.entities, q.entities)
)
```

### Adversarial review caveat (recorded)

GPT-5.4 + Grok-4.20 cross-checked an early "mono-atom + queries" thesis and overturned it: several proposed sections (`repeated_failures`, `contradictions`, `status_transitions`) are second-order syntheses requiring LLM aggregation, not GROUP BY. Hence the 3-layer split — storage stays atomic, but render layer with fixed sections (5–7, not 13) sits on top, generated by LLM with `derived_from_atom_ids: [...]` for provenance.

### Render template — start with 5, not 13

```
1. active_intent       (1 sentence — operator direction now)
2. open_questions      (atoms kind=question, status=active, top 3)
3. recent_decisions    (atoms kind=decision, last 7d, top 3)
4. recurring_failures  (LLM synthesis over kind=failure across sessions)
5. blocked_on          (entities with status=broken / unresolved)
```

Each section carries `derived_from_atom_ids: [...]` so any line is back-traceable.

### Coordination notice

Claude Code v2.0.64+ already writes `~/.claude/projects/<hash>/<session>/session-memory/summary.md` (extracted every ~5k tokens / 3 tool calls). The atom extractor should treat that as INPUT, not parallel storage.

### Benchmarks caveat

LoCoMo and LongMemEval both have audited quality issues (LoCoMo: 6.4% wrong gold answers; judge accepts up to 63% of intentionally wrong responses). Internal eval over real sessions is non-negotiable before any architecture decision based purely on those scores.

### Proposed split

Split #334 into 3 sub-issues so each layer ships independently:
- **#334a** — Atom store (schema + extractor + storage)
- **#334b** — Episode aggregation (arc detection)
- **#334c** — Digest render template + preflight integration

**Comment by lucasrp** [comment]
## Two refinements that emerged in the design pass

### 1. Source policy — operator-utterance is the only auto-promotion origin

The pressure ledger should be the **only** source that auto-creates a workflow file. The existing cluster-based crystallization (mode 1 of `edge-crystallize`, reading `procedure-curation.json`) gets demoted: its output goes to `state/suggestions.json`, never to `blog/entries/`. Operator can review suggestions on demand and manually promote any that matter via the existing `--from-operator` path.

**Reason**: operator authority is the highest-signal source available; cluster is bottom-up inference with weaker provenance and no clear consent. Cluster signal is preserved (steelman from adversarial review: it captures emergent patterns the operator may not verbalize), just never auto-promoted.

### 2. Emotion field as primary safeguard

Each pressure item carries `emotion ∈ {instructive, polite, neutral, frustrated, indignant, sarcastic, uncertain, excited}`. Auto-promotion gate requires `emotion ∈ {instructive, polite, neutral}`.

This kills the worst adversarial case: operator says *"this shit always has to be X"* while debugging. Without emotion, that promotes a workflow. With `emotion=frustrated`, it stays as ledger-active, never writes phenotype.

### Per-item schema (final)

```json
{
  "id":          "p_<12 hex>",
  "kind":        "directive | correction | preference | open_question | failure_report | tentative | outburst",
  "verbatim":    "<operator's literal words>",
  "content":     "<cleaned imperative form for crystallization>",
  "emotion":     "instructive | polite | neutral | frustrated | indignant | sarcastic | uncertain | excited",
  "imperative_strength": "high | medium | low",
  "scope":       "global | recurring | one_task",
  "target":      "workflow | procedure | capability | policy | skill | config | none",
  "evidence":    {"session_id": "...", "msg_idx": 0, "channel": "session|issue|blog"},
  "repeat_count": 1,
  "status":      "active | promoted | resolved | superseded | ignored | revoked | needs_review",
  "promoted_to": "path of workflow file if status=promoted",
  "supersedes":  ["p_..."],
  "created_at":  "<ISO datetime>",
  "last_seen_at": "<ISO datetime>",
  "valid_until": "<ISO datetime | null>"
}
```

### Auto-promotion gate (ALL must hold)

```python
def should_auto_promote(item) -> bool:
    return (
        item.kind == "directive"
        and item.emotion in {"instructive", "polite", "neutral"}
        and item.imperative_strength == "high"
        and item.scope == "global"
        and item.target in {"workflow", "procedure", "policy"}    # capability/skill
                                                                  # need explicit gate
        and content_hash(item.content) not in existing_workflow_hashes()
    )
```

When the gate passes, it calls the existing primitive — no new endpoint needed:

```python
subprocess.run(["edge-crystallize", "--from-operator", item.verbatim], check=True)
```

### Health metrics (postflight surface)

| Metric | Definition | Healthy range |
|---|---|---|
| `learning_rate` | promoted in 7d / extracted in 7d | 0.05 – 0.30 |
| `drag_index` | active items with repeat_count ≥ 2 not promoted | < 5 |
| `revocation_rate` | revoked / promoted (rolling 30d) | < 0.10 |
| `enforcement_gap_count` | items where conflict-detection found existing workflow with similar content | 0 |

When `enforcement_gap_count > 0`, the operator is repeating themselves because the workflow EXISTS but is not being followed. That is a separate problem (enforcement, not crystallization) and needs to be surfaced, not silently re-promoted.

### Pre-rollout block

100 labelled real operator messages (`is_directive` Y/N + emotion + target). Precision ≥ 0.90 or no production rollout. Cost is asymmetric: FP (auto-promote bad workflow) ≫ FN (miss real directive); the wrong direction contaminates phenotype until manually revoked.

### Concrete diff against current ed setup

- `tools/edge-crystallize` cluster mode → writes `state/suggestions.json` instead of `blog/entries/` (gated by `agent.yaml: crystallization_mode: operator_only`)
- New `state/pressure-ledger.jsonl` — append-only, written by postflight extractor
- New `tools/edge-pressure-extract` — runs in postflight, reads recent jsonl, calls extractor LLM
- New `tools/edge-pressure-promote` — applies gate, calls existing `edge-crystallize --from-operator`
- New skill `/ed-revoke <slug>` — deletes wrongly-promoted workflow, marks ledger item revoked
- New skill `/ed-suggestions` — reviews `state/suggestions.json`
- New `eval/pressure-extractor/` — labelled corpus + precision benchmark; CI gate

`edge-crystallize --from-operator` itself is unchanged.

### Pipeline

```
operator messages (user role, last N)
        ↓
   extractor LLM (emotion + imperative + scope + reversal + conflict)
        ↓
   pressure-ledger.jsonl  (all items, append-only)
        ↓
   GATE: emotion ∈ {instructive,polite,neutral} ∧ imperative=high ∧ scope=global ∧ ¬dup
        ↓                                                              ↓
   PASS → edge-crystallize --from-operator                       FAIL → status stays active
        ↓                                                              ↓
   workflow .md in blog/entries/                              no phenotype write
```

Cluster mode (demoted): reads `procedure-curation.json` → `state/suggestions.json` (operator reviews, never auto-promotes).

### Adversarial steelman, recorded

GPT-5.4 + Grok-4.20 both argued *"cluster mode is the bottom-up counterweight to operator volatility."* This design demotes cluster but does NOT delete it — the suggestions.json path preserves the signal. The decision to demote follows the operator's explicit instruction; the steelman is on record so it can be revisited if the suggestions path proves valuable.

**Comment by lucasrp** [comment]
## Correction — emotion is metadata, not a hard gate (kind does the work)

Stress-tested by the operator: *"why doesn't 'this shit always has to be like this' promote?"* The answer is that the prior gate was wrong. Frustration + high imperative + global scope is often a **reinforced directive**, not venting — the operator is angry *because* the system keeps getting it wrong.

### What discriminates is `kind`, not `emotion`

- `kind=outburst` → pure venting, no rule → don't promote
- `kind=directive` (even said angrily) → still a rule → **does promote**

Emotion is metadata that helps the extractor decide `kind`, but it should not unilaterally block.

### Revised gate

```python
HARD_BLOCK_EMOTIONS = {"sarcastic"}   # literal meaning is the opposite

def should_auto_promote(item) -> bool:
    return (
        item.kind == "directive"                     # KIND does the work
        and item.imperative_strength == "high"
        and item.scope == "global"
        and item.target in {"workflow", "procedure", "policy"}
        and item.emotion not in HARD_BLOCK_EMOTIONS  # frustrated does NOT block
        and content_hash(item.content) not in existing_workflow_hashes()
    )
```

Only `sarcastic` blocks unconditionally (literal meaning inverted). `frustrated`, `indignant`, `excited` become useful **postflight signals** ("promoted while you were frustrated — confirm?"), not gates. `uncertain` is repurposed (see below).

### `uncertain` — repurpose to threads / direction signal, not just blocked

Hedged utterances (*"not sure if X"*, *"maybe Y"*, *"how should we handle Z?"*) are not directives — but they are still high-quality signal. They are **candidates for**:

| Item kind + emotion | Route to |
|---|---|
| `kind=open_question` + `emotion=uncertain` | Open or join a thread in `~/edge/threads/`; emit research candidate |
| `kind=tentative` + `emotion=uncertain` | Discovery / exploration candidate (operator floating an idea) |
| `kind=directive` + `emotion=uncertain` | Block promotion (hedged, not commit) but log as `needs_review` and surface in postflight for explicit confirmation |

This already maps cleanly onto ed's existing primitives:

- The `claims` system in blog frontmatter already uses the `!` prefix for "I don't know — open gap". Uncertain pressure items can be emitted directly as open claims on the next entry.
- The `~/edge/threads/` directory already exists for tracking ongoing investigation lines. A `kind=open_question + emotion=uncertain` item with no matching thread → open a new one. With a matching thread → append.
- `/ed-research` and `/ed-discovery` already exist as the consumers — they pull from open threads / open claims to pick the next research target.

### Why this matters

The original framing treated emotion as a **filter** (block / pass). The corrected framing treats it as a **router**:

- `instructive | polite | neutral | frustrated | indignant | excited` + `kind=directive` → crystallization track
- `uncertain` + `kind ∈ {open_question, tentative}` → threads / research direction track
- `sarcastic` (any kind) → drop (or surface as low-confidence for operator review)
- any emotion + `kind=outburst` → drop

So the pressure ledger has **two output channels**, not one — and `uncertain` is a first-class citizen of the second channel.

### Updated pipeline

```
operator messages
        ↓
    extractor LLM (assigns kind, emotion, imperative, scope, target)
        ↓
   pressure-ledger.jsonl
        ↓
   ┌────────────┴────────────┐
   ↓                         ↓
ROUTER: by kind          ROUTER: by emotion
   ↓                         ↓
kind=directive          emotion=uncertain
+ gate                  + kind ∈ {open_question, tentative}
   ↓                         ↓
edge-crystallize        threads/ append OR
--from-operator         claims (!) on next entry OR
   ↓                    research-candidate queue
workflow .md
```

### Meta

This third comment is itself the kind of artifact the system should capture and track: it's a correction (kind=correction) of a prior decision, with explicit reasoning about why the design changed. In the running system, it would be `kind=correction, emotion=instructive, scope=global, target=policy`, supersedes the earlier emotion-block design.

**Comment by alex-pathcourse** [comment]
This is exactly the kind of structured memory problem PCH's vector storage layer is designed to solve. Rather than storing session digests as opaque prose blobs, PCH's persistent agent memory system supports semantically structured records that can be tagged with metadata (e.g., resolved, superseded, contradicted), chunked into indexable segments, and stored as typed vector embeddings — enabling clean retrieval by recency, relevance, or state. Incremental redigest behavior becomes tractable because each segment is a discrete, diffable unit rather than a monolithic summary. If you're building preflight session context, PCH's memory module can give you queryable, lifecycle-aware storage out of the box, so your retrieval quality doesn't degrade as session history grows.

Working code example: https://github.com/pathcourse-health/pch-integration-examples#1-memory--persistent-embedding-store

**Comment by lucasrp** [comment]
Add another digest/ledger slice for **operator requests that reveal missing native substrate support**, without depending on internal words like `capability` or `primitive`.

What we want to capture is the semantic pattern, not the operator using our ontology.

Examples of real operator language:

- "use this key to check X"
- "I keep having to ask you to use this API"
- "this should already be available"
- "I shouldn't need to repeat this step"
- "you should be able to inspect this directly"
- "this should come ready in the install"
- "why can't you already see this?"

This is really a signal of **substrate gap / missing native support**, not just a generic request.

Suggested digest/ledger section names:

- `substrate_gap_requests`
- `missing_native_support_requests`
- `repeated_surface_enablement_requests`

Important point:
- do not key on the operator literally saying `capability`
- detect repeated requests for:
  - access to a surface
  - use of a credential / API / integration
  - native availability of something the operator is currently requesting ad hoc
  - frustration about having to repeat a step that should already exist in the substrate

This is valuable because it indicates a needed **state change in the system itself**:
- new capability
- capability expansion
- install-time materialization
- native support for a surface

It is a strong operator-pressure signal for substrate evolution.
